src/main/java/vct/col/rewrite/SatCheckRewriter.java

```diff
@@ -13,43 +18,64 @@
+    public SatCheckRewriter(ProgramUnit source) {
+        super(source);
     }
-    Method.Kind kind=m.kind;
-    Type rt=rewrite(m.getReturnType());
-    Contract c=currentContractBuilder.getContract();
-    currentContractBuilder=null;
-    ASTNode body=rewrite(m.getBody());
-    if (body!=null && !intentional_false) switch(kind){
-    case Plain:
-    case Constructor:
-      ASTNode refute=create.special(ASTSpecial.Kind.Refute,create.constant(false));
-    	BranchOrigin branch=new BranchOrigin("Contract Unsatisfiability Check",null);
-      OriginWrapper.wrap(null,refute, branch);
-	    if (body instanceof BlockStatement){
-	    	((BlockStatement)body).prepend(refute);
-	    } else {
-	    	body=create.block(refute,body);
-	    }
-    default:
-      break;
+
+    @Override
+    public void visit(Method m) {
+        super.visit(m);
+
+        // If there is no contract it cannot be false
+        // If there is no body it's an abstract method, and hence the user is allowed to create unsoundness
+        Contract contract = m.getContract();
+        if (contract == null || m.getBody() == null) {
+            return;
+        }
+
+        // Do not sat-check if there is a false literal in the contract
+        boolean intentional_false = contract.pre_condition.isConstant(false);
+
+        if (!intentional_false && (m.kind == Method.Kind.Plain || m.kind == Method.Kind.Constructor)) {
+            ASTNode my_assert = create.special(ASTSpecial.Kind.Assert, create.constant(false));
+
+            // We want to use contract origin, but that is null at the moment (probably because of a bug)
+            // So until then we just use the method origin.
+            Origin origin = m.getOrigin();
+            if (contract.getOrigin() != null) {
+                origin = contract.getOrigin();
+            }
+            // This way we tag the assert we produced. These are then collected in SilverStatementMap.
+            // By collecting them later we allow other passes to delete them or modify them
+            // In SilverBackend we check if they indeed all error. The ones that do not error, we report as error
+            AssertOrigin my_branch = new AssertOrigin("contract satisfiability check", origin);
+
+            my_assert.clearOrigin();
+            my_assert.setOrigin(my_branch);
+
+            BlockStatement blockStatement = create.block(my_assert);
+
+            // Create method that will serve as a proof obligation for the satisfiability of the contract
+            Method assert_method = create.method_kind(
+                    m.getKind(),
+                    rewrite(m.getReturnType()),
+                    rewrite(m.getContract()),
+                    "__contract_unsatisfiable__" + m.name(),
+                    rewrite(m.getArgs()),
+                    m.usesVarArgs(),
+                    blockStatement
+            );
+
+            blockStatement.setParent(assert_method);
```

Comment:
Was this necessary? Just curious.

Reply:
Using a block statement was, but the setparent wasn't. I (randomly) noticed the field while I was debugging and thought it might be good to do. But if it isn't done consistently anywhere maybe line 75 should be removed (and possibly the parent field as well, if it isn't used anymore)